### PR TITLE
feat: detect last_changed/last_updated duration math and suggest for: field (#1157)

### LINKED ISSUE #1157: [FEATURE] Incorporate HA 2026.5 templating updates across automation, script, and template tools
### What would you like to see?

HA 2026.5 ([release notes](https://www.home-assistant.io/blog/2026/05/06/release-20265/)) ships three things worth investigating for ha-mcp — they touch the automation and script config tools as much as `ha_eval_template`:

1. **Native `for:` field on purpose-specific triggers and entity conditions** (motion, occupancy, doors, windows, lights, switches, climate, covers, media players, etc.) — replaces a large class of template-helper / `{{ now() - trigger.last_changed }}` workarounds with a first-class field. This directly affects what our `ha_config_set_automation` / `ha_config_set_script` docstrings, examples, and best-practices skill should be teaching agents.
2. **Top-to-bottom rework of the templating documentation** — 14 learning pages, a per-function reference at `/template-functions/<name>/` for all 200+ functions/filters/tests, and a dedicated error-messages page.
3. **Smart code editors with autocomplete + hover** — context-aware suggestions for functions, filters, tests, globals, and parameter completion that recognizes entity/device/area/floor/label IDs. The blog page itself ships the underlying metadata as `window.__templateFunctions` JSON (name → description, URL, parameter docs).

### Why do you need this?

We have an active line of work (#1011, PR #695) to push agents away from Jinja in trigger/condition/action positions when native alternatives exist. 2026.5 expands the "native alternative" surface significantly — our automation/script tools and the best-practices skill they lean on need to know about the new `for:` fields so they recommend the right native option, not a now-stale template workaround. The new function reference and embedded JSON also give us a structured, citable source for templating help in the places where templates *are* appropriate inside automations and scripts (dynamic service data, notification bodies, trigger context).

### Investigation areas

- **`ha_config_set_automation` / `ha_config_set_script` docstrings and examples:** review for stale "use a template helper for duration" guidance and replace with native `for:` examples. The schema goes straight to HA so the field "just works", but our examples actively teach agents the old pattern.
- **Skill / best-practices updates:** `template-guidelines.md` and `automation-patterns.md` need rows for duration-based triggers/conditions across the affected domains, and links to `/template-functions/<name>/` reference pages where they currently inline brief filter explanations.
- **Checker breadth (#1011 / PR #695):** when the "any Jinja in logic position" rewrite lands, its suggestion text for duration-style anti-patterns should point at the new native `for:` field, not just "use a state condition".
- **Template help surface for the templates we *do* allow in automations/scripts:** when an agent legitimately needs a template (dynamic service data, notification message, trigger-context access), should ha-mcp expose template-function metadata so the agent picks the right function? Options to weigh: bundle/refresh `window.__templateFunctions` JSON as a reference resource, enrich `ha_eval_template` errors with links to the per-function and error-messages pages, or add a dedicated lookup tool. Open question.

### Additional context

- HA 2026.5 release: https://www.home-assistant.io/blog/2026/05/06/release-20265/
- Template function reference URL pattern: `/template-functions/<name>/`
- The embedded JSON on the release page (`window.__templateFunctions`) maps function name → `{description, url, parameters}` for every function/filter/test
- Related: #1011, PR #695, `references/template-guidelines.md`, `references/automation-patterns.md`

### PR DESCRIPTION
**Pull request by SealKan**
## What does this PR do?

Closes #1157 (partial — covers the best-practice detector for `now() - X.last_changed/last_updated` patterns).

HA 2026.5 introduced the native `for:` field on `state` triggers and conditions for duration-based checks. This PR adds a detector in `best_practice_checker.py` that catches automations still using the old template approach (`now() - X.last_changed/last_updated`) and suggests the native `for:` alternative.

**Changes:**

- `best_practice_checker.py`: New `_RE_DURATION_MATH` regex (two alternations: `now() - X.last_changed` and `X.last_changed < now()`; both require a dotted qualifier so bare Jinja variables named `last_changed` are not falsely flagged). Warning fires in both `_check_template_string` (conditions) and `_check_triggers` (template trigger `value_template`). Warning text includes a concrete `platform: state, for: {minutes: N}` example and links to `automation-patterns.md#native-conditions` / `automation-patterns.md#trigger-types`.
- `test_best_practice_checker.py`: 8 new tests covering condition `last_changed`, condition `last_updated`, trigger template, clean positive case (`for:` field — no warning), skill-ref presence, no generic double-flag, trigger warning uses `#trigger-types` anchor, and no false positive on bare `last_changed` Jinja variable.

**Out of scope (per issue scope guidance):**
- Docstring updates to `ha_config_set_automation` / `ha_config_set_script` — separate PR once the skill content is updated upstream
- Skills repo update (`automation-patterns.md`) — separate PR to `homeassistant-ai/skills`
- Template function metadata exposure — open design question in the issue

## Type of change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 📚 Documentation
- [ ] 🔧 Maintenance/refactor
- [ ] 🧪 Tests only
- [ ] 💥 Breaking change

## Testing
- [ ] I have tested these changes with a LLM agent
- [x] All automated tests pass (`uv run pytest`)
- [x] Code follows style guidelines (`uv run ruff check`)

98 unit tests pass (was 90; 8 new). Ruff and mypy clean.

## Future improvements

- Reverse comparison `states.sensor.x.last_changed < now() - timedelta(...)` is not caught by the current regex (the `<` alternation only covers `X.last_changed < now()` with nothing following). This is a rare HA pattern; it could be added as a third alternation if it shows up in practice.
- The `numeric_state` trigger also supports a `for:` field for similar duration checks; a separate detector could be added later.

## Checklist
- [x] I have updated documentation if needed

### CODE REVIEW

File: src/ha_mcp/tools/best_practice_checker.py [hunk 1]
```diff
@@ -77,6 +77,11 @@
 _RE_STATE_IN = re.compile(r"states\s*\([^)]+\)\s+in\s+[\[(]")
 # Unsafe direct state access: states.sensor.x.state
 _RE_DIRECT_STATE = re.compile(r"\bstates\.\w+\.\w+\.state\b")
+# Duration/recency checks via last_changed or last_updated arithmetic
+_RE_DURATION_MATH = re.compile(
+    r"\bnow\(\)\s*-\s*(?:\w+\.)+last_(?:changed|updated)\b"
+    r"|\blast_(?:changed|updated)\s*<\s*now\(\)"
+)
```
[the remark this answers is not in the document]
Comment (SealKan) [reply src/ha_mcp/tools/best_practice_checker.py:86]: Fixed in commit 824d120. The second alternation now requires the same dotted qualifier as the first: ```python
_RE_DURATION_MATH = re.compile(
    r"\bnow\(\)\s*-\s*(?:\w+\.)+last_(?:changed|updated)\b"
    r"|\b(?:\w+\.)+last_(?:changed|updated)\s*<\s*now\(\)"
)
```
A bare Jinja variable named `last_changed` no longer matches. Test `test_no_false_positive_bare_last_changed_variable` covers this. The `as_timestamp(now()) - as_timestamp(...)` variant is noted as a future extension — it covers a narrower use case and would require a separate regex pass.

File: src/ha_mcp/tools/best_practice_checker.py [hunk 2]
```diff
@@ -500,6 +515,19 @@ def _check_triggers(
                             "automation-patterns.md#trigger-types",
                         )
                     )
+                if _RE_DURATION_MATH.search(vt):
```
[the remark this answers is not in the document]
Comment (SealKan) [reply src/ha_mcp/tools/best_practice_checker.py:519]: Fixed in commit 824d120. Added a dedicated `if platform == "numeric_state":` block that checks `value_template` for duration math. The block is separate from the `template` platform checks so neither the generic fallback suppression logic nor the numeric-comparison/is_state checks are affected. Test `test_numeric_state_trigger_value_template_duration_math` covers this case.

**Comment by Patch76** [comment]
RfR? I'd do some review, if you're done (for the moment).

**Comment by kingpanther13** [comment]
@SealKan please mark as ready for review if you're done working on this so we can do a formal review on it

**Comment by kingpanther13** [comment]
@SealKan can you give an update on this and your other PR please? I'm just trying to keep the pipeline moving forward.

**Comment by SealKan** [comment]
Apologies for the delay. The two gemini items were addressed in prior commits (c6a46e2 tightens the regex to require a dotted qualifier on both alternations to prevent false positives, 824d120 extends the detector to numeric_state value_template). Just pushed a ruff format fix so CI should go green. Marking ready for review now.

**Comment by SealKan** [comment]
@Patch76 following up on the Area 3 row in your May 22 status map: the Ruff Lint failure is resolved — pushed a `ruff format` pass over `best_practice_checker.py` and its test. All checks are green now, the branch is rebased on master, and the Gemini review threads are resolved, so this is no longer a draft-state item.

Per your net, that should leave only the Area 4 decision before #1157 can close — happy to defer to whatever you and the maintainers land on there. The scripts `choose`-block `for:` gap you flagged is a small docstring addition I can fold into this PR or a separate follow-up, your call.

**Review by Patch76 — CHANGES_REQUESTED** [review]
Thanks for this — the detection is well-targeted, and the test suite is genuinely thorough: condition + trigger contexts, the `for:`-already-present negative case, generic-fallback suppression, and the bare-variable false-positive guard. The regex comment matches the pattern, the warning text is accurate, and both skill-doc anchors (`#native-conditions`, `#trigger-types`) resolve.

One item before merge:

**The reversed comparison form is matched and documented, but never positively tested.** `_RE_DURATION_MATH` has two alternations:

```
now() - X.last_changed/last_updated      # forward — well covered
X.last_changed/last_updated < now()      # reversed — no positive test
```

Every positive test uses the forward form; the reversed form's only appearance is the bare variable in `test_no_false_positive_bare_last_changed_variable` (asserted *not* flagged). So the reversed alternation with a dotted qualifier — the case it exists to catch — ships without a regression guard, and a later refactor could drop it silently. Could you add one positive test per context, e.g.:

```python
"{{ states.binary_sensor.motion.last_changed < now() }}"                 # condition → expect warning
"{{ states.sensor.temp.last_updated < now() - timedelta(minutes=5) }}"   # trigger value_template → expect warning
```

One optional, non-blocking note: the duration assertions match on substring (any warning containing the fragments), so asserting exactly one such warning fires would also catch accidental double-flagging.

The reversed-form tests are the only blocker here; the rest is solid.

**Comment by SealKan** [comment]
Good catch @Patch76 - you're right, the reversed form had no positive coverage. Added one test per context (condition + trigger value_template) for the dotted-qualifier `X.last_changed/last_updated < now()` case in 5177d96. I also took your double-flagging note and asserted exactly one duration warning fires in each, rather than just "a warning containing the fragments", so an accidental double-flag would now fail too. Both forward and reversed alternations are guarded now.

**Review by Patch76 — COMMENTED** [review]
Thanks for this — the detector is clean and the test coverage is genuinely thorough (both attributes, both regex alternations, the `numeric_state` path, the bare-`last_changed` false-positive guard, and double-flag suppression are all covered). The condition→`#native-conditions` / trigger→`#trigger-types` anchor split is correct and explicitly tested. None of the following are blockers.

**1. PR description is out of date — the reversed form *is* caught.**
The "Future improvements" note says `states.sensor.x.last_changed < now() - timedelta(...)` isn't matched by the current regex. It actually is: the second alternation `(?:\w+\.)+last_(?:changed|updated)\s*<\s*now\(\)` matches `X.last_changed < now()` and doesn't care what follows `now()`, so the offset form matches too. Your own `test_trigger_reversed_comparison_last_updated_math` and `test_condition_reversed_comparison_last_changed_math` confirm this. Worth deleting that bullet so the description doesn't contradict the tests.

**2. `numeric_state` path has no anchor assertion.**
`test_numeric_state_trigger_value_template_duration_math` asserts on the message text (`"last_changed/last_updated"`, `"for:"`) but not on the `#trigger-types` anchor, so a copy-paste slip to `#native-conditions` in the `numeric_state` block (`best_practice_checker.py` ~L549) wouldn't be caught — unlike the template-trigger path, which has `test_trigger_warning_uses_trigger_types_anchor`. A one-line addition mirroring that test (run with `skill_prefix=SKILL_PREFIX`, assert `automation-patterns.md#trigger-types`) closes the gap.

**3. Nit — regex comment wording.**
`best_practice_checker.py:81` says "at least one dotted qualifier". `(?:\w+\.)+` is one-*or-more*, so `trigger.last_changed` (one dot) and `states.sensor.x.last_changed` (three) both match. "one or more dotted qualifiers" reads more precisely. Trivial.

For the record: I checked the new regex for ReDoS. It's technically O(n²) on a pathological all-dots input (a property it shares with the other dotted-path regexes in this file), but at realistic template depths it stays well under a millisecond, so there's nothing to do here.

**Review by Patch76 — APPROVED** [review]
The reversed-comparison regression tests I asked for are now in place — `test_condition_reversed_comparison_last_changed_math` and `test_trigger_reversed_comparison_last_updated_math`, both asserting exactly one warning fires, which also closes the double-flag concern. That was the only blocker, so I am approving.

The three items in my companion comment (the stale "Future improvements" bullet, the missing `#trigger-types` anchor assertion on the `numeric_state` test, and the comment-wording nit) are all non-blocking — nice to fold in while you are here, but none gate merge. Thanks for the thorough tests.
